### PR TITLE
[aggregator] Handle follower canLead() for not yet flushed windows

### PR DESCRIPTION
**Pull request by abliqo**
### Problem

Recently we started using status endpoint to track health of aggregator leaders and followers.
We noticed that some followers are occasionally in a state of `canLead == false` when the last flush time recorded by leader is zero. Specifically, it was noticed for forwarded metrics at 10 minute resolution.

According to the leader flush manager code, this is possible when a new metric at some resolution appears for the very first time for a given shard (a new list is created). But since leader persists flush times for all resolution widows in the same batch, if some resolution window is brand new, its last flush time in etcd is zero.

Follower does not handle the zero value of last flush time and reports `canLead == false` even when it can legitimately takeover leadership.

### Solution

When follower encounters a resolution window with last flush time equal zero, it should report `canLead == true` if it has been up for amount of time greater than that resolution window. 

**Does this PR introduce a user-facing and/or backwards incompatible change?**:
NONE

**Does this PR require updating code package or user-facing documentation?**:
NONE

### CODE REVIEW

File: src/aggregator/aggregator/follower_flush_mgr.go [hunk 1]
```diff
@@ -267,7 +281,21 @@ func (mgr *followerFlushManager) CanLead() bool {
 			if waitTillFlushedTime.Equal(mgr.openedAt) {
 				waitTillFlushedTime = waitTillFlushedTime.Add(-windowSize)
 			}
-			for _, lastFlushedNanos := range fbr.ByNumForwardedTimes {
+
+			for numForwardedTimes, lastFlushedNanos := range fbr.ByNumForwardedTimes {
+				if lastFlushedNanos == 0 {
+					mgr.logger.Warn("Encountered a window that was never flushed by leader",
+						zap.String("windowSize", windowSize.String()),
```
Comment (vdarulis) [inline src/aggregator/aggregator/follower_flush_mgr.go:288]: zap.Stringer

File: src/aggregator/aggregator/follower_flush_mgr.go [hunk 2]
```diff
@@ -279,12 +307,39 @@ func (mgr *followerFlushManager) CanLead() bool {
...
+	flusherType string,
+	shardID int,
 	flushTimes map[int64]int64,
 	metrics standardFollowerFlusherMetrics,
 ) bool {
 	for windowNanos, lastFlushedNanos := range flushTimes {
 		windowSize := time.Duration(windowNanos)
+		if lastFlushedNanos == 0 {
+			mgr.logger.Warn("Encountered a window that was never flushed by leader",
+				zap.String("windowSize", windowSize.String()),
```
Comment (vdarulis) [inline src/aggregator/aggregator/follower_flush_mgr.go:333]: zap.Stringer

File: src/aggregator/aggregator/follower_flush_mgr.go [hunk 3]
```diff
@@ -279,12 +307,39 @@ func (mgr *followerFlushManager) CanLead() bool {
...
+	earliestPossibleWindowStart := mgr.nowFn().Add(-windowSize)
+	if mgr.openedAt.Before(earliestPossibleWindowStart) {
+		return true
+	}
+
+	return false
+}
+
 func (mgr *followerFlushManager) canLead(
+	flusherType string,
```
Comment (vdarulis) [inline src/aggregator/aggregator/follower_flush_mgr.go:324]: make this a const enum generated w/ stringer?

File: src/aggregator/aggregator/follower_flush_mgr.go [hunk 4]
```diff
@@ -243,17 +247,27 @@ func (mgr *followerFlushManager) CanLead() bool {
...
+			mgr.metrics.timed) {
 			return false
 		}
 
 		// Check that the forwarded metrics have been flushed past the process start
 		// time, meaning the forwarded metrics that didn't make to the process have been
 		// flushed successfully downstream.
 		for windowNanos, fbr := range shardFlushTimes.ForwardedByResolution {
+			mgr.logger.Warn("ForwardedByResolution is nil",
+				zap.Int("shardID", int(shardID)))
```
Comment (vdarulis) [inline src/aggregator/aggregator/follower_flush_mgr.go:270]: should be in if statement below?

File: src/aggregator/aggregator/follower_flush_mgr.go [hunk 5]
```diff
@@ -279,12 +307,39 @@ func (mgr *followerFlushManager) CanLead() bool {
 	return true
 }
 
+// canLeadNotFlushed determines whether the follower can takeover leadership
+// for the window that was not (yet) flushed by leader.
+// This is case is possible when leader encounters a metric at some resolution
+// window for the first time in the shard it owns.
+func (mgr *followerFlushManager) canLeadNotFushed(windowSize time.Duration) bool {
+	earliestPossibleWindowStart := mgr.nowFn().Add(-windowSize)
```
Comment (andrewmains12) [inline src/aggregator/aggregator/follower_flush_mgr.go:315]: I think we might want `Truncate` here instead, to put this at the actual window start. Also, don't we still need the same logic for `openedAt` that we use in the other cases?

File: src/aggregator/aggregator/follower_flush_mgr.go [hunk 6]
```diff
@@ -243,17 +247,27 @@ func (mgr *followerFlushManager) CanLead() bool {
 		// Check that for standard metrics, all the open windows containing the process
 		// start time are closed, meaning the standard metrics that didn't make to the
 		// process have been flushed successfully downstream.
-		if !mgr.canLead(shardFlushTimes.StandardByResolution, mgr.metrics.standard) {
+		if !mgr.canLead(
+			"standard",
+			int(shardID),
+			shardFlushTimes.StandardByResolution,
+			mgr.metrics.standard) {
```
Comment (vdarulis) [inline src/aggregator/aggregator/follower_flush_mgr.go:254]: add a newline before `) {`

File: src/aggregator/aggregator/follower_flush_mgr_test.go [hunk 7]
```diff
@@ -131,6 +132,56 @@ func TestFollowerFlushManagerCanNotLeadForwardedFlushWindowsNotEnded(t *testing.
 	require.False(t, mgr.CanLead())
 }
 
+func TestFollowerFlushManagerCanLeadNotFlushed(t *testing.T) {
```
Comment (andrewmains12) [inline src/aggregator/aggregator/follower_flush_mgr_test.go:135]: Can we add some test cases around window boundaries? Things like:

```
manager opened right after window start
manager opened right before window start
```
etc.

File: src/aggregator/aggregator/follower_flush_mgr.go (hunk 3 above)
Comment (andrewmains12) [reply src/aggregator/aggregator/follower_flush_mgr.go:324]: Good call--could use  http://github.com/m3db/m3/blob/cb2d5f7ad7b426fa48315b2a946d6127cf61774e/src/aggregator/aggregator/list.go#L766-L766

File: src/aggregator/aggregator/follower_flush_mgr.go [hunk 8]
```diff
@@ -267,7 +281,21 @@ func (mgr *followerFlushManager) CanLead() bool {
 			if waitTillFlushedTime.Equal(mgr.openedAt) {
 				waitTillFlushedTime = waitTillFlushedTime.Add(-windowSize)
 			}
-			for _, lastFlushedNanos := range fbr.ByNumForwardedTimes {
+
+			for numForwardedTimes, lastFlushedNanos := range fbr.ByNumForwardedTimes {
+				if lastFlushedNanos == 0 {
+					mgr.logger.Warn("Encountered a window that was never flushed by leader",
```
Comment (andrewmains12) [inline src/aggregator/aggregator/follower_flush_mgr.go:287]: Maybe should just be `Info`? It's not unexpected.

File: src/aggregator/aggregator/follower_flush_mgr.go (hunk 4 above)
Comment (abliqo) [reply src/aggregator/aggregator/follower_flush_mgr.go:270]: oops, yes

File: src/aggregator/aggregator/follower_flush_mgr_test.go [hunk 9]
```diff
@@ -131,6 +132,56 @@ func TestFollowerFlushManagerCanNotLeadForwardedFlushWindowsNotEnded(t *testing.
...
+						},
+					},
+				},
+			},
+		},
+	}
+
+	mgr.processed = testFlushTimes
+
+	mgr.openedAt = time.Now().Add(-9 * time.Minute)
```
Comment (andrewmains12) [inline src/aggregator/aggregator/follower_flush_mgr_test.go:178]: Nit: let's do these as separate test cases (could table test it) instead of multiple asserts in one case.

File: src/aggregator/aggregator/follower_flush_mgr.go (hunk 5 above)
Comment (abliqo) [reply src/aggregator/aggregator/follower_flush_mgr.go:315]: Agree about `Truncate`!
Since we don't know when the never-flushed window started, we assume that it started at the most recent possible time, so our best effort is to use "now" to determine its start time.

File: src/aggregator/aggregator/follower_flush_mgr_test.go (hunk 7 above)
Comment (abliqo) [reply src/aggregator/aggregator/follower_flush_mgr_test.go:135]: Both cases in the bottom of this test.

File: src/aggregator/aggregator/follower_flush_mgr.go (hunk 8 above)
Comment (abliqo) [reply src/aggregator/aggregator/follower_flush_mgr.go:287]: fair!

File: src/aggregator/aggregator/follower_flush_mgr.go (hunk 5 above)
Comment (andrewmains12) [reply src/aggregator/aggregator/follower_flush_mgr.go:315]: > Since we don't know when the never-flushed window started, we assume that it started at the most recent possible time, so our best effort is to use "now" to determine its start time.

Ahh ok I think I get what you're saying. So basically in the normal case, we find "nearest window to the open time" and compare that to the flush, whereas here we're saying "make sure you were up when the nearest window to now started". Seems right I think!

Comment (andrewmains12) [inline src/aggregator/aggregator/follower_flush_mgr.go:315]: Oh btw, let's use a consistent `now` for this--we can call `nowFn` at the top and then pass it through.

Comment (abliqo) [reply src/aggregator/aggregator/follower_flush_mgr.go:315]: Well, it would make code chattier, because `mgr.nowFn()` is not used in the callsites of `canLeadNotFlushed()`. But `mgr.nowFn()` is consistently used in other methods not related to `canLead`.

File: src/aggregator/aggregator/follower_flush_mgr_test.go (hunk 9 above)
Comment (abliqo) [reply src/aggregator/aggregator/follower_flush_mgr_test.go:178]: Table tests are less obvious to read and debug.
I encapsulated a test logic in a function and added multiple `t.Run()` as separate test cases.

File: src/aggregator/aggregator/follower_flush_mgr.go (hunk 5 above)
Comment (andrewmains12) [reply src/aggregator/aggregator/follower_flush_mgr.go]: Yeah not a lot for it I think; I think it's still worth getting a consistent snapshot if we can (obviously only need to pass it through where we need it).

Comment (abliqo) [reply src/aggregator/aggregator/follower_flush_mgr.go]: Oh got it. Done, thanks!

File: src/aggregator/aggregator/follower_flush_mgr.go [hunk 10]
```diff
@@ -267,7 +286,21 @@ func (mgr *followerFlushManager) CanLead() bool {
 			if waitTillFlushedTime.Equal(mgr.openedAt) {
 				waitTillFlushedTime = waitTillFlushedTime.Add(-windowSize)
 			}
-			for _, lastFlushedNanos := range fbr.ByNumForwardedTimes {
+
+			for numForwardedTimes, lastFlushedNanos := range fbr.ByNumForwardedTimes {
+				if lastFlushedNanos == 0 {
+					mgr.logger.Info("Encountered a window that was never flushed by leader",
+						zap.Stringer("windowSize", windowSize),
```
Comment (andrewmains12) [inline src/aggregator/aggregator/follower_flush_mgr.go:293]: Nit: let's log `now` as well? Might as well, and I think with that we can fully determine what's going on.

File: src/aggregator/aggregator/follower_flush_mgr.go [hunk 11]
```diff
@@ -267,7 +286,21 @@ func (mgr *followerFlushManager) CanLead() bool {
...
+						zap.Stringer("flusherType", forwardedMetricListType),
+						zap.Int("shardID", int(shardID)),
+						zap.Int32("numForwardedTimes", numForwardedTimes))
+					mgr.metrics.forwarded.windowNeverFlushed.Inc(1)
+
+					if mgr.canLeadNotFushed(now, windowSize) {
```
Comment (andrewmains12) [inline src/aggregator/aggregator/follower_flush_mgr.go:299]: Spelling nit: `s/canLeadNotFushed/canLeadNotFlushed/g`

File: src/aggregator/aggregator/follower_flush_mgr.go [hunk 12]
```diff
@@ -279,12 +312,36 @@ func (mgr *followerFlushManager) CanLead() bool {
...
 	for windowNanos, lastFlushedNanos := range flushTimes {
 		windowSize := time.Duration(windowNanos)
+		if lastFlushedNanos == 0 {
+			mgr.logger.Info("Encountered a window that was never flushed by leader",
+				zap.Stringer("windowSize", windowSize),
+				zap.Stringer("flusherType", flusherType),
+				zap.Int("shardID", int(shardID)))
+			mgr.metrics.forwarded.windowNeverFlushed.Inc(1)
+
+			if mgr.canLeadNotFushed(now, windowSize) {
```
Comment (andrewmains12) [inline src/aggregator/aggregator/follower_flush_mgr.go:340]: Nit: maybe more consistent with the other code (below) to do:

```
if !mgr.canLeadNotFlushed(...) {
   return false
}
```

Either way though.

File: src/aggregator/aggregator/follower_flush_mgr_test.go [hunk 13]
```diff
@@ -131,6 +132,63 @@ func TestFollowerFlushManagerCanNotLeadForwardedFlushWindowsNotEnded(t *testing.
...
+		opts := NewFlushManagerOptions().SetElectionManager(electionManager)
+		mgr := newFollowerFlushManager(doneCh, opts).(*followerFlushManager)
+
+		mgr.processed = testFlushTimes
+		mgr.openedAt = followerOpenedAt
+		require.Equal(t, expectedCanLead, mgr.CanLead())
+	}
+
+	t.Run("opened_on_the_window_start", func(t *testing.T) {
+		followerOpenedAt := time.Now().Truncate(window)
```
Comment (andrewmains12) [inline src/aggregator/aggregator/follower_flush_mgr_test.go:174]: These tests don't flake do they? Not sure if worth it, but we could use a mocked out time for this to make it super clear what times are when (`time.Now()` is a bit more opaque). Either way though.

**Review by andrewmains12 — APPROVED** [review]

File: src/aggregator/aggregator/follower_flush_mgr.go (hunk 12 above)
Comment (abliqo) [reply src/aggregator/aggregator/follower_flush_mgr.go:340]: If it's a never flushed window and `canLeadNotFlushed(...)` is `true` for it, we have to `continue` to skip the further 
check that assumes the window is flushed window. But if it's false then yes, need to return early. So it actually has to be like this:

```
if mgr.canLeadNotFlushed(now, windowSize) {
    continue
} else {
    return false
}
```

File: src/aggregator/aggregator/follower_flush_mgr_test.go (hunk 13 above)
Comment (abliqo) [reply src/aggregator/aggregator/follower_flush_mgr_test.go:174]: They are not flaky, but I agree, for clarity it's better to pass the mocked fixed now.

File: src/aggregator/aggregator/follower_flush_mgr.go (hunk 12 above)
Comment (andrewmains12) [reply src/aggregator/aggregator/follower_flush_mgr.go]: Ah yeah I see your point, good catch. Do we have test cases around the `return false` case?

**Review by andrewmains12 — APPROVED** [review]
